### PR TITLE
Phase 6: Test-Time Augmentation via AoA Perturbation

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The `aug="aoa_perturb"` training augmentation applies random ±1° AoA rotations during training to improve generalization. But at inference, the model receives a single input — no test-time augmentation is applied.

**Test-Time Augmentation (TTA)** evaluates the model at multiple slightly perturbed inputs and averages the predictions. For CFD surrogates, the natural perturbation is angle of attack: the pressure distribution at AoA=α is smooth and similar to the solution at AoA=α±δ.

**Physical motivation:** CFD solutions are smooth functions of AoA. A neural surrogate with residual overfitting will produce slightly different error patterns at AoA ± δ. Averaging these predictions cancels error components that vary rapidly with AoA (model artifacts) while preserving the true physical signal (which varies slowly).

**Key advantage: ZERO training cost.** This is an inference-only technique that can be applied to any existing checkpoint. If it works, it "stacks" with any future training improvement. 3× inference cost is the tradeoff.

**Never tested at inference time in this programme.** The `aoa_perturb` flag only exists for training.

## Instructions

**Step 1: Add TTA flags** to argparse:
```python
parser.add_argument('--tta_aoa', action='store_true', default=False,
                    help='Enable TTA via AoA perturbation at inference')
parser.add_argument('--tta_aoa_delta', type=float, default=1.0,
                    help='AoA perturbation magnitude in degrees')
```
Add to Config dataclass:
```python
tta_aoa: bool = False
tta_aoa_delta: float = 1.0
```

**Step 2: Implement TTA in the evaluation loop.** Find where the model is evaluated on validation splits. Apply AoA perturbation (coordinate rotation) to the input, run the model, then **rotate velocity predictions back** before averaging.

```python
if cfg.tta_aoa:
    aoa_deltas = [-cfg.tta_aoa_delta, 0.0, cfg.tta_aoa_delta]  # 3 views
    preds_list = []
    
    for delta in aoa_deltas:
        x_aug = x.clone()
        if abs(delta) > 1e-6:
            angle_rad = delta * (torch.pi / 180.0)
            cos_a = torch.cos(torch.tensor(angle_rad, device=x.device))
            sin_a = torch.sin(torch.tensor(angle_rad, device=x.device))
            # Rotate spatial coordinates (x[:,:,0:2])
            x_orig = x_aug[:, :, 0:1].clone()
            y_orig = x_aug[:, :, 1:2].clone()
            x_aug[:, :, 0:1] = cos_a * x_orig - sin_a * y_orig
            x_aug[:, :, 1:2] = sin_a * x_orig + cos_a * y_orig
        
        with torch.no_grad():
            pred = model(x_aug, ...)["preds"]
            # CRITICAL: Rotate velocity predictions BACK to original frame
            if abs(delta) > 1e-6:
                ux = pred[:, :, 0:1].clone()
                uy = pred[:, :, 1:2].clone()
                pred[:, :, 0:1] = cos_a * ux + sin_a * uy    # inverse rotation
                pred[:, :, 1:2] = -sin_a * ux + cos_a * uy
                # Pressure (channel 2) is scalar — no rotation needed
        preds_list.append(pred)
    
    final_pred = torch.stack(preds_list).mean(0)
else:
    final_pred = model(x, ...)["preds"]
```

**CRITICAL: Velocity inverse rotation.** The velocity predictions (Ux, Uy) are vectors and MUST be rotated back from the perturbed frame to the original frame before averaging. Pressure is a scalar and needs no rotation. Failing to unrotate will corrupt Ux/Uy predictions.

**Step 3: Find where `pos` / spatial coordinates are in `x`.** Look at the existing `aoa_perturb` training augmentation code to see which channels are the (x, y) coordinates. The TTA must apply the same rotation.

**Step 4: Also check if DSDF features need rotation.** The `aug_full_dsdf_rot` flag suggests that DSDF gradient features are also rotated during training augmentation. If the eval-time TTA rotates coordinates, it should also rotate DSDF features consistently. Check the training `aoa_perturb` implementation for which features get rotated.

**Experiment approach:** Train a baseline model, then evaluate it WITH and WITHOUT TTA. No retraining needed.

```bash
# Step 1: Train baseline (or use existing checkpoints)
python train.py --agent alphonse \
  --wandb_name "alphonse/tta-baseline-s42" \
  --wandb_group phase6/tta-aoa \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3

# Step 2: Evaluate the SAME checkpoint with TTA
python train.py --agent alphonse \
  --wandb_name "alphonse/tta-delta1-s42" \
  --wandb_group phase6/tta-aoa \
  --tta_aoa --tta_aoa_delta 1.0 \
  --seed 42 [same flags] --eval_only  # if eval_only flag exists

# Step 3: Try different deltas
python train.py ... --tta_aoa --tta_aoa_delta 0.5 ...   # ±0.5°
python train.py ... --tta_aoa --tta_aoa_delta 2.0 ...   # ±2.0°
```

**If there's no `--eval_only` flag**, you'll need to add the TTA logic to the end-of-training evaluation block. In that case, train 4 runs: 2 seeds × (baseline eval vs TTA eval), and use the same checkpoint for both by saving and reloading.

**Alternative approach if eval_only is hard:** Implement TTA directly in the existing evaluation block that runs at the end of training. Add the `--tta_aoa` flag and apply TTA during the final validation evaluation. Then run 2 baseline + 2 TTA-enabled training runs (same seeds) and compare the final evaluation metrics.

W&B group: `phase6/tta-aoa`

## Baseline

Current best (16-seed ensemble, PR #2093):

| Metric | Baseline |
|--------|----------|
| p_in   | **12.1** |
| p_oodc | **6.6**  |
| p_tan  | **29.1** |
| p_re   | **5.8**  |

Single-model references:
- seed=42: p_in≈12.71, p_oodc≈8.13
- seed=73: p_in≈12.2, p_oodc≈7.59

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse \
  --wandb_name "alphonse/baseline-s42" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

All 8 runs hit the 180-min wall-clock limit. **Critical finding: TTA validation overhead causes fewer training epochs, confounding the comparison.**

### Surface MAE — Pressure (mae_surf_p, lower=better)

| Run | W&B ID | Seed | Best Epoch | p_in | p_tan | p_oodc | p_re | val/loss |
|-----|--------|------|-----------|------|-------|--------|------|----------|
| baseline | m6p2jskq | 42 | 157 | 12.9 | 30.4 | 7.9 | 6.5 | 0.3906 |
| baseline | u24b9oh9 | 43 | 157 | 12.9 | 29.8 | 7.8 | 6.4 | 0.3844 |
| tta-d0.5 | 6biat14l | 42 | 130 | 16.3 | 30.5 | 8.8 | 7.0 | 0.4169 |
| tta-d0.5 | j604eyv0 | 43 | 132 | 14.1 | 29.8 | 8.7 | 6.9 | 0.4040 |
| tta-d1.0 | j7n5d3wf | 42 | 132 | 13.7 | 31.4 | 8.5 | 7.1 | 0.4114 |
| tta-d1.0 | z4kv5ghh | 43 | 131 | 14.3 | 31.2 | 8.5 | 6.9 | 0.4091 |
| tta-d2.0 | b6uyjpm6 | 42 | 131 | 16.2 | 30.6 | 8.8 | 7.4 | 0.4187 |
| tta-d2.0 | n7bs7408 | 43 | 131 | 14.5 | 31.3 | 8.6 | 7.0 | 0.4091 |

### Training Identity Verification

**Confirmed: training is identical at matching epochs.** I compared epoch-by-epoch val/loss between baseline and TTA runs (same seed). At every epoch, the train metrics and val/loss are bit-for-bit identical:

```
Epoch 120: baseline train[vol=0.0850 surf=0.0110] val[in=0.3031 tan=0.9476 oodc=0.2353 re=0.2231]
Epoch 120: tta-d1.0 train[vol=0.0850 surf=0.0110] val[in=0.3031 tan=0.9476 oodc=0.2353 re=0.2231]
```

This confirms TTA does NOT corrupt training or val/loss computation. The only effect is on mae_surf_p (physical-space metrics via pred_orig averaging).

### The Confound: Validation Overhead

| Config | Epoch time | Max epoch reached | Overhead |
|--------|-----------|-------------------|----------|
| baseline | 68s | 157 | — |
| tta (any delta) | 82s | 130-132 | +14s/epoch (+21%) |

TTA runs 3 forward passes per validation batch (−δ, 0, +δ), adding ~14s per epoch. Within the 180-min timeout, TTA runs train for **25 fewer epochs** (130 vs 157). Since the model was still improving at epoch 130, the mae_surf_p comparison is confounded by training length.

### What This Means

The apparent degradation in TTA runs is **primarily due to 25 fewer training epochs**, not TTA averaging itself. I cannot isolate the pure TTA effect from this experiment because:
1. No `--eval_only` flag exists to evaluate a trained checkpoint with/without TTA
2. The mae_surf_p at matching epochs is not logged to output.log (only val/loss is)
3. W&B should have per-epoch mae_surf_p — the advisor may be able to extract this from the W&B dashboard

### Commands Used

```bash
BASE="--agent alphonse --wandb_group phase6/tta-aoa \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3"

# GPU 0-1: baseline (seeds 42, 43)
# GPU 2-3: --tta_aoa --tta_aoa_delta 0.5 (seeds 42, 43)
# GPU 4-5: --tta_aoa --tta_aoa_delta 1.0 (seeds 42, 43)
# GPU 6-7: --tta_aoa --tta_aoa_delta 2.0 (seeds 42, 43)
```

### Peak Memory
38.0 GB (baseline), 38.6 GB (TTA variants)

### What Happened

TTA is implemented correctly (training is verified identical, velocity predictions are properly inverse-rotated before averaging). However, the 3x validation cost confounds the experiment within our 180-min timeout constraint. The raw numbers show TTA runs are worse, but this is primarily due to fewer training epochs rather than TTA averaging being harmful.

### Suggested Follow-ups

1. **Add `--eval_only` flag** to load a trained checkpoint and evaluate with/without TTA. This would cleanly isolate the TTA effect by comparing the same model with and without TTA averaging. This is the proper way to evaluate TTA.
2. **Extract per-epoch mae_surf_p from W&B** for baseline and TTA runs at matching epochs (e.g., epoch 130). If mae_surf_p is logged per epoch, this would reveal whether TTA helps, hurts, or is neutral at the same training stage.
3. **Apply TTA post-hoc to the 8/16-seed ensemble checkpoints.** If TTA shows any benefit per-model, it stacks multiplicatively with ensembling.
4. **Consider TTA only at final eval** (not every validation epoch) to avoid the training length confound while still fitting within the 180-min timeout.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2111 because ADVISOR: Thorough analysis — closing as a marginally negative/neutral result.

**W&B epoch-matched analysis (at equal training length, epoch 130):**

| Group | p_in | p_tan | p_oodc | p_re |
|-------|------|-------|--------|------|
| Baseline (2-seed avg) | 15.71 | 31.43 | 8.88 | 7.20 |
| TTA d=0.5 (2-seed avg) | 15.71 | 31.30 | 8.81 | 7.15 |
| TTA d=1.0 (2-seed avg) | 15.88 | 31.29 | 8.80 | 7.13 |

**Delta vs baseline at epoch 130 (negative = improvement):**
- TTA d=0.5: p_in ≈ -0.0%, p_tan -0.4%, p_oodc -0.8%, p_re -0.8%
- TTA d=1.0: p_in +1.1%, p_tan -0.4%, p_oodc -0.9%, p_re -1.0%

**Analysis:** You correctly diagnosed the confound — the apparent degradation in final metrics is entirely explained by 25 fewer training epochs (those 27 extra epochs the baseline gets are worth ~18% improvement on p_in alone). At matching training length, TTA provides mild signal on p_tan/p_oodc/p_re (~0.4-1.0%) but this is:

1. **Below statistical significance** with only 2 seeds per condition
2. **Self-defeating under our timeout constraint:** the 3x validation overhead that buys you ~0.8% gain also costs you 25 training epochs worth ~18% loss — a net very negative tradeoff
3. **Inference-cost prohibitive:** 3x inference overhead per model × 16 ensemble models = 48x total inference cost for < 1% gain

**Conclusion:** TTA as a training-time validation strategy is not viable under our 180-min timeout. Your diagnosis was scientifically excellent and the implementation was correct. The core issue is structural: any inference-augmentation technique that runs at every validation epoch destroys the training budget.

**Your suggestion #3 (apply TTA post-hoc to existing ensemble checkpoints)** is the one worth considering independently — it costs nothing during training. If we apply TTA to each of the 16 ensemble models and re-ensemble, the variance reduction might compound. But the epoch-matched data suggests the per-model gain is too small (≤1%) to prioritize over training improvements. Filing this as a known marginal technique rather than a dead end.

Great work on the thorough diagnosis. Moving you to a new direction with higher expected value..
